File: web/src/entities/service/model/diagnosis.ts
import type { Diagnosis } from "@/shared/api";

export const DIAGNOSIS_MESSAGES = {
  "local-network-denied": { title: "diagnosis.local-network-denied.title", action: "diagnosis.local-network-denied.action" },
  refused: { title: "diagnosis.refused.title", action: "diagnosis.refused.action" },
  timeout: { title: "diagnosis.timeout.title", action: "diagnosis.timeout.action" },
  "host-unreachable": { title: "diagnosis.host-unreachable.title", action: "diagnosis.host-unreachable.action" },
  "name-not-resolved": { title: "diagnosis.name-not-resolved.title", action: "diagnosis.name-not-resolved.action" },
  tls: { title: "diagnosis.tls.title", action: "diagnosis.tls.action" },
  "not-http": { title: "diagnosis.not-http.title", action: "diagnosis.not-http.action" },
  "http-status": { title: "diagnosis.http-status.title", action: "diagnosis.http-status.action" },
  "icmp-not-permitted": { title: "diagnosis.icmp-not-permitted.title", action: "diagnosis.icmp-not-permitted.action" },
  other: { title: "diagnosis.other.title", action: "diagnosis.other.action" },
} as const satisfies Record<Diagnosis, { title: string; action: string }>;

export function diagnosisMessage(code: Diagnosis | null) {
  return code ? DIAGNOSIS_MESSAGES[code] : null;
}

File: crates/core/portal-model/src/types/diagnosis.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Diagnosis {
    LocalNetworkDenied,
    Refused,
    Timeout,
    HostUnreachable,
    NameNotResolved,
    Tls,
    NotHttp,
    HttpStatus,
    IcmpNotPermitted,
    Other,
    #[serde(other)]
    Unknown,
}

impl Diagnosis {
    pub const ALL: [Diagnosis; 10] = [
        Diagnosis::LocalNetworkDenied,
        Diagnosis::Refused,
        Diagnosis::Timeout,
        Diagnosis::HostUnreachable,
        Diagnosis::NameNotResolved,
        Diagnosis::Tls,
        Diagnosis::NotHttp,
        Diagnosis::HttpStatus,
        Diagnosis::IcmpNotPermitted,
        Diagnosis::Other,
    ];

    pub fn code(self) -> &'static str {
        match self {
            Diagnosis::LocalNetworkDenied => "local-network-denied",
            Diagnosis::Refused => "refused",
            Diagnosis::Timeout => "timeout",
            Diagnosis::HostUnreachable => "host-unreachable",
            Diagnosis::NameNotResolved => "name-not-resolved",
            Diagnosis::Tls => "tls",
            Diagnosis::NotHttp => "not-http",
            Diagnosis::HttpStatus => "http-status",
            Diagnosis::IcmpNotPermitted => "icmp-not-permitted",
            Diagnosis::Other => "other",
            Diagnosis::Unknown => "unknown",
        }
    }
}
